### PR TITLE
renameutils: fix package on darwin

### PR DESCRIPTION
**Pull request by arulagrawal**
###### Description of changes

Fixed the compilation of `renameutils` on aarch64 darwin, and since `renameutils` requires gnu `ls`, `mv`, `cp` at runtime, I have adapted a [patch](https://github.com/Homebrew/homebrew-core/blob/d46782e45325badc160c3f2bdf0381b64714c7b0/Formula/renameutils.rb#L37) from homebrew,  and a postPatch to put the correct paths in place.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/misc/renameutils/default.nix [hunk 1]
```diff
@@ -9,9 +9,25 @@ stdenv.mkDerivation rec {
     sha256 = "18xlkr56jdyajjihcmfqlyyanzyiqqlzbhrm6695mkvw081g1lnb";
   };
 
-  patches = [ ./install-exec.patch ];
+  patches = [ ./install-exec.patch ] ++ lib.optionals stdenv.isDarwin [ ./darwin.patch ];
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/renameutils/default.nix:12]: Optional patches are not great. Also we do not use the g prefix for coreutils normally. Please try to wrap instead. Also why do we need to patch this first an later replace?

File: pkgs/tools/misc/renameutils/install-exec.patch [hunk 2]
```diff
@@ -3,7 +3,7 @@ diff -Naur renameutils-0.12.0-orig/src/Makefile.am renameutils-0.12.0/src/Makefi
 +++ renameutils-0.12.0/src/Makefile.am	2013-03-25 11:13:21.605447377 +0000
 @@ -49,7 +49,7 @@
  	@[ -f icp ] || (echo $(LN_S) icmd icp ; $(LN_S) icmd icp)
- 
+
  install-exec-local:
```
[the remark this answers is not in the document]
Comment (arulagrawal) [reply pkgs/tools/misc/renameutils/install-exec.patch:7]: Yep, my bad, looks like my editor had its own opinions about whitespace. I will remove this file from the PR.

File: pkgs/tools/misc/renameutils/default.nix (hunk 1 above)
Comment (arulagrawal) [reply pkgs/tools/misc/renameutils/default.nix]: I have removed the patch in favour of doing the substitutions directly in the postpatch substituteInPlace.

File: pkgs/tools/misc/renameutils/default.nix [hunk 3]
```diff
@@ -11,7 +11,23 @@ stdenv.mkDerivation rec {
 
   patches = [ ./install-exec.patch ];
 
+  preConfigure = lib.optionalString (stdenv.isDarwin && stdenv.isAarch64) ''
+    export ac_cv_func_lstat64=no
+  '';
+
   nativeBuildInputs = [ readline ];
+  buildInputs = lib.optionals stdenv.isDarwin [ coreutils ];
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/renameutils/default.nix:19]:
```suggestion
```
This won't do anything if coreutils is only used in postPatch

File: pkgs/tools/misc/renameutils/default.nix [hunk 4]
```diff
@@ -11,8 +11,23 @@ stdenv.mkDerivation rec {
...
+  postPatch = lib.optionalString stdenv.isDarwin ''
+    substituteInPlace src/apply.c \
+      --replace "command = \"mv\"" "command = \"${coreutils}/bin/mv\"" \
+      --replace "command = \"cp\"" "command = \"${coreutils}/bin/cp\""
+    substituteInPlace src/icmd.c \
+      --replace "#define MV_COMMAND \"mv\"" "#define MV_COMMAND \"${coreutils}/bin/mv\"" \
+      --replace "#define CP_COMMAND \"cp\"" "#define CP_COMMAND \"${coreutils}/bin/cp\""
+    substituteInPlace src/qcmd.c \
+      --replace "ls_program = xstrdup(\"ls\")" "ls_program = xstrdup(\"${coreutils}/bin/ls\")"
+  '';
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/renameutils/default.nix:29]:
```suggestion
  patches = [ ./install-exec.patch ];

  postPatch = lib.optionalString stdenv.isDarwin ''
    substituteInPlace src/apply.c \
      --replace "command = \"mv\"" "command = \"${coreutils}/bin/mv\"" \
      --replace "command = \"cp\"" "command = \"${coreutils}/bin/cp\""
    substituteInPlace src/icmd.c \
      --replace "#define MV_COMMAND \"mv\"" "#define MV_COMMAND \"${coreutils}/bin/mv\"" \
      --replace "#define CP_COMMAND \"cp\"" "#define CP_COMMAND \"${coreutils}/bin/cp\""
    substituteInPlace src/qcmd.c \
      --replace "ls_program = xstrdup(\"ls\")" "ls_program = xstrdup(\"${coreutils}/bin/ls\")"
  '';

  nativeBuildInputs = [ readline ];

  preConfigure = lib.optionalString (stdenv.isDarwin && stdenv.isAarch64) ''
    export ac_cv_func_lstat64=no
  '';
```

**Pull request merged**